Added "add custom node" functionality

gemdev111 opened the pull request:
<table>
  <tr>
    <td><img src="https://github.com/gemwalletcom/gem-ios/assets/171273137/fd3f7a93-29cd-4db2-9154-96bd33858ba6" style="width: 45%;" /></td>
    <td><img src="https://github.com/gemwalletcom/gem-ios/assets/171273137/17c43b30-fa74-416c-b11a-57b842e64935" style="width: 45%;" /></td>
  </tr>
</table>

gemdev111 committed:
Refactored blockchain services

Added MARKs, now they are more clear
Added latest block for eth provider, Added ChainLatestBlockFetchable for services

gemdev111 committed:
Added ImportNodeScene & ImportNodeSceneViewModel with logic to find and import custom node #9

Added  import single node function to ImportNodeService
Added presentation from ChainSettingsScene
Minor refactored ChainSettingsScene
Updated Localizations

gemdev111 committed:
added 1 more case for disabled state, if not isInSync - disable

gemcoder21 commented on Gem/Nodes/Scenes/ImportNodeScene.swift:
```diff
@@ -0,0 +1,162 @@
+
+    private func onSelectImport() {
+        model.importFoundNode()
+        onDismiss()
```
dismiss should be called once? `onDismiss` and handeled from the outside of this screen

gemdev111 replied:
make sence, will do

gemcoder21 commented on Gem/Nodes/ViewModels/ImportNodeSceneViewModel.swift:
```diff
@@ -0,0 +1,139 @@
+
+    @Published var urlString: String = ""
+    @Published var state: StateViewType<Bool> = .noData
+    @Published var chainID: String = ""
```
Instead of having chainID, block number and in sync variables, maybe create `ImportNodeResult` that will have those values ?

gemcoder21 commented on Gem/Nodes/ViewModels/ImportNodeSceneViewModel.swift:
```diff
@@ -0,0 +1,139 @@
+extension ImportNodeSceneViewModel {
+    func importFoundNode()  {
+        let node = Node(url: urlString, status: .active, priority: 10)
+        try? importNodeService.importNode(ChainNodes(chain: chain.rawValue, nodes: [node]))
```
we might need to error an error here, avoid using `try?`

gemcoder21 commented on Gem/Nodes/ViewModels/ImportNodeSceneViewModel.swift:
```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    private func getPrimaryNetworkId(for chain: Chain) -> String {
```
use `StakeConfig.config()` it will return config for each chain that will contain `network_id`

gemcoder21 replied:
if `network_id` optional, do not check network id field

gemcoder21 commented on Gem/Resources/zh-Hant.lproj/Localizable.strings:
```diff
@@ -200,4 +200,12 @@
-"social.facebook" = "Facebook";
\ No newline at end of file
+"social.facebook" = "Facebook";
+"settings.networks.import_node.title" = "Add node";
```
I would recommend to simplify it: `nodes.import_node`

gemcoder21 commented on Packages/Blockchain/Sources/Ethereum/EthereumService.swift:
```diff
@@ -287,3 +306,14 @@ extension EthereumService: ChainIDFetchable {
+// MARK: - ChainLatestBlockFetchable
+
+extension EthereumService: ChainLatestBlockFetchable {
+    public func getLatestBlock() async throws -> String? {
```
```suggestion
    public func getLatestBlock() async throws -> String {
```

String should be non optional

gemdev111 committed:
Updates (ref to @gemcoder21 reivew )

Added ImportNodeResult instead of simple properties, intergrated it into state object
Removed optional try?
Updated localizations
Removed unecessary dismiss() on onSelectImport
Added sheet for QR scanner
Updated get getLatestBlock in services and in ImportNodeResult to non optional
Updated validation "chainID" logic by using ChainConfig

gemcoder21 commented on Gem/Nodes/Scenes/ImportNodeScene.swift:
```diff
@@ -0,0 +1,171 @@
+// MARK: - UI Components
+
+extension ImportNodeScene {
+    private var inputSection: some View {
```
can you create an issue to create this custom view? it's used in many places in the app, need to have reusability, recipient, custom address, here

gemdev111 replied:
https://github.com/orgs/gemwalletcom/projects/2/views/4?pane=issue&itemId=67986270

gemdev111 committed:
Minor Updates for "add custom node"

 Simply added "URL" title instead of "RPC Address" for input field
Added formatter for block number
Localization Updates
